### PR TITLE
C#: Document generated members

### LINKED ISSUE #65996: C# generated code does not include public XML comments or disable 1591
### Godot version

4.0.0-beta1

### System information

macOS 12

### Issue description

Running `dotnet build` in C# Projects which have `<GenerateDocumentationFile>true</GenerateDocumentationFile>` specified in their `*.csproj` files results in warnings from the code produced by the Godot dotnet source generators:

```
warning CS1591: Missing XML comment for publicly visible type or member 'Main.SignalName'
warning CS1591: Missing XML comment for publicly visible type or member 'Main.RestoreGodotObjectData(GodotSerializationInfo)'
...etc
```

I currently maintain 5 packages for Godot whose public API's are fully documented and require `GenerateDocumentationFile` on those packages.

Possible solution:

The Godot source generators are already emitting `#pragma warning disable` for other rules (here's [one such example](https://github.com/godotengine/godot/blob/master/modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptMethodsGenerator.cs#L136)).

Since Microsoft [doesn't seem to care](https://github.com/dotnet/roslyn/issues/12702#issuecomment-285483035), I believe we should add a `#pragma warning disable 1591` to every new or modified API added by the Godot dotnet source generators.

### Steps to reproduce

Essentially, any project which contains `<GenerateDocumentationFile>true</GenerateDocumentationFile>` in the `*.csproj` file will result in build warnings about missing documentation from the Godot dotnet source generator output.

### Minimal reproduction project

I am working on updating one of my smallest packages to Godot 4. You can reference the `godot4` branch of `go_dot_collections` here as a minimum reproduction sample: https://github.com/chickensoft-games/go_dot_collections/tree/godot4

### PR DESCRIPTION
**Pull request by raulsntos**
Documents generated members and tries to discourage users from calling/overriding internal methods that only exist to be used by the engine.

- Fixes https://github.com/godotengine/godot/issues/65996.

### CODE REVIEW

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptMethodsGenerator.cs [hunk 1]
```diff
@@ -147,6 +151,12 @@ INamedTypeSymbol symbol
 
             foreach (string methodName in distinctMethodNames)
             {
+                source.Append("        /// <summary>\n")
+                    .Append("        /// Cached name for the '")
+                    .Append(methodName)
+                    .Append("' method.\n")
```
Comment (RedworkDE) [inline modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptMethodsGenerator.cs:157]:
```suggestion
                    .Append($"        /// Cached StringName for the '{methodName}' method.\n")
```

or even

```suggestion
                    .Append($"        /// Cached <see cref=\"Godot.StringName\" /> for the <see cref=\"{methodName}\" /> method.\n")
```

(same for the other such cases)

**Review by RedworkDE — COMMENTED** [review]
> tries to discourage users from calling/overriding internal methods that only exist to be used by the engine.

Should even properly document them in xml docs then or should we just have the warning not to use them and whatever is is necessary to suppress the warnings?

Related: net8 adds an `Experimental` attribute for things like this: https://github.com/dotnet/designs/blob/main/accepted/2023/preview-apis/preview-apis.md which we could use for these methods once that is released. Or alternatively we could mark these things `Obsolete` to generate a warning right now when using them.

**Comment by raulsntos** [comment]
> net8 adds an `Experimental` attribute for things like this: [dotnet/designs@main/accepted/2023/preview-apis/preview-apis.md](https://github.com/dotnet/designs/blob/main/accepted/2023/preview-apis/preview-apis.md?rgh-link-date=2023-07-10T07%3A39%3A05Z) which we could use for these methods once that is released

I don't feel like the `Experimental` attribute fits these APIs, since they are not preview features.

> Or alternatively we could mark these things `Obsolete` to generate a warning right now when using them.

Some of these methods are used by source generators so this would generate warnings for users that don't use them directly. Also, it would generate warnings for us in the GodotSharp project.

> Should even properly document them in xml docs then or should we just have the warning not to use them and whatever is is necessary to suppress the warnings?

I think documenting them is useful, because it can be helpful for contributors. I don't think it's that important for users not to call these methods, just letting them know that they are not mean to and hiding them so they don't accidentally stumble upon them is enough.

File: modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptMethodsGenerator.cs (hunk 1 above)
Comment (RedworkDE) [reply modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptMethodsGenerator.cs:157]: I think the better readability is worth the performance impact, but it doesn't really matter too much either way so keeping it consistent with the remaining code is probably better.

**Comment by RedworkDE** [comment]
With the `Obsolete` / `Experimental` I was mostly thinking about future changes´to these methods, eg we explicitly and clearly signal that we may end up changing these methods if future improvements require it. (Unless this is interpretations is completely wrong and they are supposed to be proper members of the public interface, and if it was possible to make them inaccessible they would still be public)

**Comment by raulsntos** [comment]
My concern is that since some of these methods are used by source generators, those warnings would show up in user projects even if they are not using them directly.

Also, technically we should avoid breaking compat in these methods too, because an external library built against Godot 4.0 could be using them. For example, if a library implements a Node type that the user derives from in their game project, then I think the generated `InvokeGodotClassMethod` method will throw a `MissingMethodException` when calling the base method implementation.

And I see what you mean now by using `Obsolete` / `Experimental` to mark that these methods may change in the future, but I still feel it's weird to mark the API as obsolete or experimental for this. It's just a matter of semantics for me.

**Review by RedworkDE — APPROVED** [review]

**Pull request merged**